File: Binary Trees/2_depth.py
# Height of Binary Tree (Max Depth)
def height(root):
    if root is None:
        return 0
    
    left = height(root.left)
    right = height(root.right)

    ans = max(left, right) + 1
    return ans



# Min Depth of Binary Tree
def minDepth(root) :
    if root is None:
        return 0
    
    left = minDepth(root.left)
    right = minDepth(root.right)

    if left == 0:
        return right + 1
    if right == 0:
        return left + 1

    ans = min(left, right) + 1
    return ans



# Given a Binary Tree, check if all leaves are at same level or not.
def height(root): 
    if root is None:
        return 0
    
    left = height(root.left)
    right = height(root.right)

    ans = max(left, right) + 1
    return ans

def minDepth(root):
    if root is None:
        return 0
    
    left  = minDepth(root.left)
    right = minDepth(root.right)

    if left == 0:
        return right + 1
    if right == 0:
        return left + 1
    
    ans = min(left, right) + 1
    return ans

def check(root):
    max_depth = height(root)
    min_depth = minDepth(root)

    return max_depth == min_depth




# 814. Binary Tree Pruning
def pruneTree(root):
    if root is None:
        return None

    root.left = pruneTree(root.left)
    root.right = pruneTree(root.right)

    if(root.left is None and root.right is None and root.val == 0):
        return None
    
    return root

File: Binary Trees cpp/2_depth.cpp
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode(int val) {
        this->val = val;
        this->left = NULL;
        this->right = NULL;
    }
}


// 104. Height of Binary Tree (Max Depth)
int height(TreeNode* root) {
    if (root == nullptr) {
        return 0;
    }

    int left = height(root->left);
    int right = height(root->right);

    int ans = max(left, right) + 1;
    return ans;
}



// 111. Min Depth of Binary Tree
int minDepth(TreeNode* root) {
    if (root == nullptr) {
        return 0;
    }

    int left = minDepth(root->left);
    int right = minDepth(root->right);

    if (left == 0) {
        return right + 1;
    }
    if (right == 0) {
        return left + 1;
    }

    int ans = min(left, right) + 1;
    return ans;
}



// Leaf Nodes at Same Level 
bool check(TreeNode* root) {
    int max_depth = height(root);
    int min_depth = minDepth(root);

    return max_depth == min_depth;
}



// 814. Binary Tree Pruning
TreeNode* pruneTree(TreeNode* root) {
    if (root == nullptr) {
        return nullptr;
    }

    root->left = pruneTree(root->left);
    root->right = pruneTree(root->right);

    if (root->left == nullptr && root->right == nullptr && root->val == 0) {
        return nullptr;
    }

    return root;
}


// Duplicate subtree in Binary Tree
// 652. Find Duplicate Subtrees 